File: Cses/LIS.py
n=int(input())
a=list(map(int, input().split()))

dp=[1]*(n+1)
dp[0]=1
ans=0
for i in range(1, n):
    for j in range(i-1, -1, -1):
        if(a[j]<a[i]):
            dp[i]=max(dp[i], 1+dp[j])
    ans=max(ans, dp[i])
print(ans)    

File: Cses/LIS.cpp

#include <bits/stdc++.h>
using namespace std;
typedef long long ll;

int main()
{
    int n;
    cin >> n;
    vector<int> a(n);
    for (int i = 0; i < n; i++)
    {
        cin >> a[i];
    }
    ll ans = 1;
    vector<ll> dp(n + 1, 1);
    for (int i = n - 1; i >= 0; i--)
    {
        for (int j = i + 1; j < n; j++)
        {
            if (a[j] > a[i])
            {
                dp[i] = max(dp[i], 1 + dp[j]);
            }
        }
        ans = max(ans, dp[i]);
    }
    cout << ans;
    return 0;
}